lib/srv/discovery/discovery.go

```diff
@@ -1330,8 +1322,91 @@ func (s *Server) getAllGCPServerFetchers() []server.Fetcher {
 	return allFetchers
 }
 
+// acquireDiscoveryGroup tries to acquire a lock if the Service has a DiscoveryGroup
+// It will retry using an exponential backoff algorithm.
+func (s *Server) acquireDiscoveryGroup() error {
+	if s.DiscoveryGroup == "" {
+		s.Log.Warnf("DiscoveryGroup is not set, skipping semaphore lock. It is recommended to set a DiscoveryGroup")
+		return nil
+	}
+
+	retry, err := retryutils.NewRetryV2(retryutils.RetryV2Config{
+		First:  0,
+		Driver: retryutils.NewExponentialDriver(defaults.HighResPollingPeriod),
+		Max:    defaults.LowResPollingPeriod,
+		Jitter: retryutils.NewHalfJitter(),
+		Clock:  s.clock,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for range retry.After() {
+		retry.Inc()
+		s.Log.Debugf("Discovery service is trying to acquire lock for DiscoveryGroup %q", s.DiscoveryGroup)
+		if err := s.tryAcquireDiscoveryGroupLease(); err != nil {
+			if strings.Contains(err.Error(), teleport.MaxLeases) {
+				s.Log.Debugf("Discovery service is waiting on DiscoveryGroup %q lock: %v", s.DiscoveryGroup, err)
+				continue
+			}
+			return trace.Wrap(err)
+		}
+		break
+	}
+
+	return nil
+}
+
+// tryAcquireDiscoveryGroupLease tries to acquire a semaphore lock on the DiscoveryService/<DiscoveryGroup> key.
+// If successful, a new goroutine is created to handle lease termination.
+// A lease can be terminated in two ways:
+// - Server's context is cancelled
+// - lease is lost
+//
+// When the lease is lost, the server is stopped.
+// Calling Stop twice is safe because it is protected by a `sync.Once`.
+// The 2nd time it is called, it will be a no-op.
+func (s *Server) tryAcquireDiscoveryGroupLease() error {
+	lease, err := services.AcquireSemaphoreLock(s.ctx, services.SemaphoreLockConfig{
+		Service: s.Config.AccessPoint,
+		Expiry:  time.Minute,
+		Params: types.AcquireSemaphoreRequest{
+			SemaphoreKind: types.SemaphoreKindDiscoveryServiceGroup,
+			SemaphoreName: s.DiscoveryGroup,
+			MaxLeases:     1,
+			Holder:        s.ServerID,
+		},
+		Clock: s.clock,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		for {
+			select {
+			case <-lease.Renewed():
+				continue
+
+			// A lease.Done() signal means that the lease was lost.
+			// The server must be stopped to ensure only one DiscoveryService is running at the same time.
+			case <-lease.Done():
+				s.Log.WithError(lease.Wait()).Warnf("DiscoveryGroup %q lock was lost, stopping discovery service.", s.DiscoveryGroup)
+				s.Stop()
```

Comment:
I don't think you can call s.Stop here.
There are several reasons of it:
- Once you loose a lock you can regain it after a while. This happens several times during auth server restarts where auth is restarted and long-lived connections are droped and the agent loses the lease although the lock is still in its "name". Once auth is back alive the agent can resume the connection. 
- This process is only valid if you have multiple discovery agents running. If you just have one it will cause the discovery service do stop after the first auth disconnection and it won't reconnect
- Stopping the service will cause issues to `/healthz` report as the critical service discovery service won't be running anymore and teleport will report unhealthy. This causes kubernetes to restart pods and load balancers to stop sending traffic.

Reply:
Yeah, you are correct.
I'll remove the semaphore lock logic and rely entirely on server name to ensure we don't duplicate the nodes.

I still think we should invest in preventing duplicate API calls, but I'll leave that to a future PR.